### ISSUE TITLE
Better explanation of the Vanadium context in tutorials

### ISSUE THREAD

**Issue by yyoon** [vanadium/issues#1142]
@asadovsky @jxson

I was chatting with @cosnicolaou about my previous onboarding experience with Vanadium, and I remembered that the tutorials on v.io didn't explain anything about Vanadium context.

Vanadium context is probably one of the first things programmers would interact with, and we should explain it better in the tutorials.

**Comment by jxson** [comment]
Really good point /cc @aghassemi

**Comment by aghassemi** [comment]
I agree. My opinion has been that developers should not need to interact with things until they actually need to use them. We force developers to pass context around without really knowing (yet) what it is for so we can tie our own stuff together. I understand that changing this behaviour (by maybe having an implicit singleton root context and having ctx be an optional arg for when developer actually needs a new context) may not be possible in all languages but I would prefer such approach much more than starting a "hello world" tutorial by explaining what context is.

@mattr-google

**Comment by ghost** [comment]
@yyoon Yes we should have better documentation for sure.

@aghassemi Your approach fundamentally breaks what the context is trying to do.  If people don't add it from the beginning the chances of ever getting meaningful traces is very low.  Libraries will evolve to ignore context, and it's extremely difficult to address after you have a program running.

Having a global singleton is simply not useful, we may as well not have the context at all.  Your contention that most applications don't need a new context is highly suspect.  The most important thing is for servers to use the context they get on a call for further outgoing calls.

Now if  you want to argue that the context, tracing, passing language etc. around just aren't worth it, OK.  We can debate whether we should totally eliminate the context, though in my opinion the time for that debate has passed.

**Comment by AsturaPhoenix** [comment]
The Baku Toolkit in Java attempts to address this issue by taking an OO approach to context management, where wrapper classes have their own Vanadium contexts and pass them around as appropriate and allowing modification where desirable. Contexts are thus bound to semantic task units, facilitating lifecycle management while allowing most developers to disregard them unless they need different timeouts or something.

I believe we can do the same thing in Go. It is probably still worthwhile to keep a simple and easily composeable base library but also provide utility/helper/wrapper libraries that attempt to organize things into larger common components.

**Comment by jxson** [comment]
@aghassemi 

> I would prefer such approach much more than starting a "hello world" tutorial by explaining what context is.

I agree that a detailed explanation would be wrong in that particular context (no pun intended).

There are a lot of details that become necessary once you pass that first tutorial stage though. Ideally in a "Hello World" scenario we could very briefly mention what the `context` is and why its necessary while providing a link to deeper documentation. Anyone retracing their steps would rediscover the link and dive deeper if needed or drop a link to the explanation in their own code comments etc.

**Comment by ghost** [comment]
I don't fully understand what you mean, but embedding contexts in other objects is an anti-pattern.  The reason is then it becomes difficult to decide if you have a context available or not.  It works best when the context is just the first param to every function.

**Comment by ghost** [comment]
This is good background reading:

https://blog.golang.org/context

**Comment by AsturaPhoenix** [comment]
The background content mostly pertains to request handling though; for client code it's a lot less applicable. I see this in my code as well; on the server-style handlers I fully appreciate the context, but for client-style logic I only use it for timeouts/cancellation and, indirectly, instrumentation.

I don't understand what you mean by it being difficult to decide if you have a context available or not if the context is included in an object. Can you give an example?

For an example of what I mean, the Baku Android Toolkit includes a Vanadium Context in its Syncbase client, app, db, and table ReactiveX wrappers. Each step in the hierarchy inherits its parent context, and each wrapper method uses its object's context by default. In this way, client code can simply call, for example, table.get(key) if they don't need to override anything.

**Comment by cosnicolaou** [comment]
We're not going to change our APIs now, for better or worse, it is what it
is. I resisted a 'forced 1st parameter' for the longest time, but the more
apps and tests we built the more I realised that my resistance was utterly
futile. Without the context many things become very difficult/impossible.
Examples include:
- being sure that all library code can be run with a different identity
  which is essential for building shared services or tools, ditto for
  logging. Think of it as enabling composition.
- providing a mechanism for configuring, controlling nested RPCs and
  libraries, collecting debugging information and tracing RPCs across process
  boundaries.

basically, consistency is a good thing, even it's overhead for simple use
cases. You can build any simple thing any way you want, what I wanted to
enable was creating complex apps out of simple components and to have those
components work well together.

The tutorial would start off by saying 'for this simple example you can
ignore the context, but you'll learn to love it when you write real
applications and libraries that work consistently with each other'. The
context would be discussed in more detail in the security tutorial which is
the first place it starts to add some value.

Cheers, Cos.

**Comment by aghassemi** [comment]
I am not arguing against usefulness of context, I am arguing that the "if you don't do this from the start, you will regret it later" part is not a strong enough case to justify enforcement of it. Not being able to trace nested RPCs is something that 1- Some developers may never encounter 2- When they do, they will go and add it (sure it is a bit late, but, whatever, they missed one chance to debug), but at that point they know what context is and why it is useful. Think of a similar scenario: I can argue that if developers are not logging enough information, they will regret it later, which is fair point but would it be okay if I (somehow) force every method to have a ".log()" call and fail to build if there isn't one? Maybe a better approach is to inject the log calls for them or give them warning instead of failing.

**Comment by ghost** [comment]
Yeah, I'll try to write up a better explanation of why the context is the way it is.  I've done a bad job at this.

@aghassemi Adding logging later is really easy.  Adding context is REALLY hard.  You have to change your function signature and the signature of EVERY function above you in the call chain.  Basically adding context after the fact means completely changing all your APIs.  It's virtually impossible.  I don't think your analogy is good.

**Comment by cosnicolaou** [comment]
On Thu, Feb 11, 2016 at 5:23 PM, Matt Rosencrantz notifications@github.com
wrote:

> Yeah, I'll try to write up a better explanation of why the context is the
> way it is. I've done a bad job at this.
> 
> @aghassemi https://github.com/aghassemi Adding logging later is really
> easy. Adding context is REALLY hard. You have to change your function
> signature and the signature of EVERY function above you in the call chain.
> Basically adding context after the fact means completely changing all your
> APIs. It's virtually impossible. I don't think your analogy is good.

I made a lot of the changes to add the context and can attest that it was
very, very, arduous a year ago and today would be impossible. This is
absolutely an all or nothing thing. You can't add 'tracing rpcs' since it
would require changing code that you don't control, so again, this is
something we very explicitly decided to design in from the get-go. If
developers don't need rpc and don't need security then they don't need
vanadium - and I have no problem with that all!

Cheers, Cos.

**Comment by aghassemi** [comment]
Fair points. I think this sort of explanation (specifically the part about it being hard to be added after the fact based on real experience) goes a long way justifying why context is not optional and should be included somewhere in the documentation of context.
